stage0/src/zero_page.rs

```diff
@@ -59,6 +79,39 @@ pub fn init_zero_page(fw_cfg: &mut FwCfg) -> &'static mut BootParams {
     zero_page
 }
 
+/// Builds an E820 table by reading the low and high memory amount from CMOS.
+///
+/// The code is largely based on what SeaBIOS is doing, but
+/// <https://wiki.osdev.org/Detecting_Memory_%28x86%29> is also a good read on the topic.
+fn build_e820_from_nvram(
+    zero_page: &mut BootParams,
+    port_factory: PortFactoryWrapper,
+) -> Result<(), &'static str> {
+    // Safety: fw_cfg is available, so we're running under QEMU(ish), and QEMU has CMOS available at
+    // the required I/O ports.
```

Comment:
Is this always the case in QEMU? i.e. does it not depend on the machine type, for instance?

Reply:
Maybe. We're only using `microvm` machine type, and that has a pre-populated E820 table available, so we won't hit this code path under QEMU. With a certain other QEMU-compatible VMM, however, we do need to use this interface.

I've expanded a comment a bit to hedge some of the risks. All in all, like I said in the comment above, we're just playing the same game as SeaBIOS here.